new group deferal strategy and collision handling

gfr10598 opened the pull request:
Previously, c2s and s2c tests were processed and forwarded to the inserter as soon as a meta file was available.  If a new test came in before .meta file, the tests were processed without the meta file.

This lead to orphaned tests if the tests at the end of a tar file had no meta file.

This restructures the handling of test groups, and allows explicitly processing the last group at the end of the tar file, in the Flush() function.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/etl/127)
<!-- Reviewable:end -->

pboothe commented on parser/ndt.go:
```diff
@@ -149,22 +151,24 @@ func (n *NDTParser) ParseAndInsert(taskInfo map[string]bigquery.Value, testName
+	// Because of port number, the c2s, s2c, and meta files may come in
+	// any order.  We defer processing until Flush or new test group.
+	// TODO - should we just ignore non-gzipped test files?  Or do some archives
+	// have unzipped files?
```
It looks like gzip compression capabilities did not arrive until after MLab started ( https://github.com/ndt-project/ndt/blob/master/src/logging.c#L45 ). So I can confidently say that ignoring all non-zipped results is the wrong thing.  Ignoring unzipped files if the zipped ones are present is definitely the right thing.

pboothe commented on parser/ndt.go:
```diff
@@ -177,11 +181,10 @@ func (n *NDTParser) ParseAndInsert(taskInfo map[string]bigquery.Value, testName
-			n.processTest(taskFileName, n.c2s, "c2s")
 		}
+		// We always use the latest file, since .gz is more reliably
+		// complete, and lexicographically later.
```
Okay, but then this ups the priority on `TODO(prod): make sure historical tar files are also in lexicographic order`

gfr10598 commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[parser/ndt.go, line 171 at r1](https://reviewable.io:443/reviews/m-lab/etl/127#-KmRs5UmJWIwaRD-vBp2-r1-171:-KmSDOtOkFTSqNf2x4oR:bxgmu3e) ([raw file](https://github.com/m-lab/etl/blob/9ac7ed9b5186df024e75f92a33617e32fd1d1a1c/parser/ndt.go#L171)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary><blockquote>

It looks like gzip compression capabilities did not arrive until after MLab started ( https://github.com/ndt-project/ndt/blob/master/src/logging.c#L45 ). So I can confidently say that ignoring all non-zipped results is the wrong thing.  Ignoring unzipped files if the zipped ones are present is definitely the right thing.
</blockquote></details>

Updated.  Also tweaked code below.

---

*[parser/ndt.go, line 186 at r1](https://reviewable.io:443/reviews/m-lab/etl/127#-KmRs5UmJWIwaRD-vBp2-r1-186:-KmSDVppaSruK5UgCnJY:bjdf5b7) ([raw file](https://github.com/m-lab/etl/blob/9ac7ed9b5186df024e75f92a33617e32fd1d1a1c/parser/ndt.go#L186)):*
<details><summary><i>Previously, pboothe (Peter Boothe) wrote…</i></summary><blockquote>

Okay, but then this ups the priority on `TODO(prod): make sure historical tar files are also in lexicographic order`
</blockquote></details>

Updated to actually prefer the .gz explicitly.
Added check to ensure we handle non-lexical ordering cases, and to detect and report timestamps out of order.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/127)*
<!-- Sent from Reviewable.io -->

gfr10598 commented:
PTAL

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/etl/127#-:-KmSFpoPSe2SFk_HgscC:bb71pro)*
<!-- Sent from Reviewable.io -->

pboothe commented:
LGTM
